Sources/TSCBasic/Path.swift

```diff
@@ -179,24 +255,50 @@ public struct AbsolutePath: Hashable {
     /// NOTE: We may want to consider adding operators such as `+` for appending
     ///       a path component.
 
-    /// NOTE: We will want to add a method to return the lowest common ancestor
-    ///       path.
-
-    /// Root directory (whose string representation is just a path separator).
-    public static let root = AbsolutePath(PathImpl.root)
+    /// Returns the lowest common ancestor path.
+    public func lowestCommonAncestor(with path: AbsolutePath) -> AbsolutePath? {
+        guard root == path.root else {
+            return nil
+        }
+        var filepath = path.filepath
+        while (!filepath.isRoot) {
+            if self.filepath.starts(with: filepath) {
+                break
+            }
+            filepath.removeLastComponent()
+        }
+        return AbsolutePath(filepath)
+    }
 
-    /// Normalized string representation (the normalization rules are described
-    /// in the documentation of the initializer).  This string is never empty.
-    public var pathString: String {
-        return _impl.string
+    /// The root directory. It is always `/` on UNIX, but may vary on Windows.
+    @available(*, deprecated, message: "root is not a static value, use the instance property instead")
+    public static var root: AbsolutePath {
+        if let rootPath = localFileSystem.currentWorkingDirectory?.root {
+            return AbsolutePath(rootPath)
+        } else {
+#if !os(Windows)
+            return AbsolutePath("/")
+#else
+            if let drive = ProcessEnv.vars["SystemDrive"] ?? ProcessEnv.vars["HomeDrive"] {
+                return AbsolutePath(drive + "\\")
+            } else {
+                fatalError("cannot determine the drive")
+            }
+#endif
+        }
     }
 
-    /// Returns an array of strings that make up the path components of the
-    /// absolute path.  This is the same sequence of strings as the basenames
-    /// of each successive path component, starting from the root.  Therefore
-    /// the first path component of an absolute path is always `/`.
-    public var components: [String] {
-        return _impl.components
+    public static func withPOSIX(path: String) -> AbsolutePath {
+#if os(Windows)
+        var filepath = FilePath(path)
+        precondition(filepath.root != nil)
+        if !filepath.isAbsolute {
+            filepath.root = root.filepath.root
+        }
+        return AbsolutePath(filepath)
```

Comment:
`public static func withPOSIX(path: String) -> AbsolutePath` is intended for using POSIX-like file systems that have a single root on Windows. Therefore, the `path` should always begin with `/`, then implicitly transformed to `\`, which is a valid root for Windows.

Correspondingly, if `path` doesn’t begin with `/`, `AbsolutePath(path)` will have precondition failures on macOS and Linux.

Reply:
It would be better if we have `PosixFilePath` and `WindowsFilePath`, with conditional `typealias FilePath` to the default one. POSIX file paths are useful sometimes on Windows.